server/src/test/java/org/apache/kafka/server/ClientMetricsManagerTest.java

```diff
@@ -919,4 +922,102 @@ public void testPushTelemetryConcurrentRequestAfterSubscriptionUpdate() throws U
         // 1 request should fail with throttling error.
         assertEquals(1, throttlingErrorCount);
     }
+
+    @Test
+    public void testCacheEviction() throws UnknownHostException, InterruptedException {
+        Properties properties = new Properties();
+        properties.put("metrics", ClientMetricsConfigs.ALL_SUBSCRIBED_METRICS_CONFIG);
+        properties.put(ClientMetricsConfigs.PUSH_INTERVAL_MS, "100");
+        clientMetricsManager.updateSubscription("sub-1", properties);
+
+        GetTelemetrySubscriptionsRequest request = new GetTelemetrySubscriptionsRequest.Builder(
+            new GetTelemetrySubscriptionsRequestData(), true).build();
+
+        GetTelemetrySubscriptionsResponse response = clientMetricsManager.processGetTelemetrySubscriptionRequest(
+            request, ClientMetricsTestUtils.requestContext());
+        assertEquals(Errors.NONE, response.error());
+
+        assertNotNull(clientMetricsManager.clientInstance(response.data().clientInstanceId()));
+        assertEquals(1, clientMetricsManager.expirationTimer().size());
+        // Cache expiry should occur after 100 * 3 = 300 ms, wait for at most 600 ms for the eviction
+        // to happen as eviction timer is scheduled in different thread.
+        assertTimeoutPreemptively(Duration.ofMillis(600), () -> {
+            // Validate that cache eviction happens and client instance is removed from cache.
+            while (clientMetricsManager.expirationTimer().size() != 0 ||
```

Comment:
The is a busy loop and is not CPU friendly. Could we use sth like `TestUtils.waitUntilTrue` instead? TestUtils is in `core`. We could probably move it to `server-common`.

Reply:
I checked the code for `waitUntilTrue` which actually takes a pause with thread sleep. Though I agree the current code has a busy wait but refactoring the method from TestUtils.scala to server-common with changes in all test files seems to be huge. Rather I could create a TestUtils java class in server-common with `waitUntilTrue` method as like scala and could use that.
But I have also done equivalent now in the current code where thread sleeps for 100 ms in the busy wait. Also the executable in `assertTimeoutPreemptively` is executed in different thread. Please let me know if it works.

Reply:
There is a flakiness if clock is not forced to advance, hence I have added that back and lowered the max wait time in assertion.